오픈마켓 I [STEP 2] 주디, 재재

ZZBAE opened the pull request:
@protocorn93 콘 안녕하세요!
step2의 PR 보냅니다. 리뷰 잘 부탁드리겠습니다 🙇🏻‍♀️


# 🛍 오픈마켓 OPEN MARKET
> **프로젝트 기간** : 2022.07.13 수  ~ 2022.07.20 수 </br>**팀원** : 주디, 재재</br>**리뷰어** : 콘

## 💾 개발환경 및 라이브러리
[![swift](https://img.shields.io/badge/swift-5.6-orange)]()
[![xcode](https://img.shields.io/badge/Xcode-13.4.1-blue)]()

---

## 💡 Step 2 수행 중 핵심경험
- [x] Safe Area을 고려한 오토 레이아웃 구현
- [x] Collection View의 활용
- [x] Mordern Collection View 활용
<br>

## 📱 구현한 화면
- 로딩화면 - 리스트화면 - 그리드화면
![](https://i.imgur.com/uuLB7At.gif)

<br>

## 🧑‍💻 전반적인 코드 설명

### **Model** 
#### 1) **Item**
> product Model에서 가져온 정보를 뷰에다가 뿌려줄 Item 정보들
- `productImage`(상품이미지), `productName`(상품이름), `price`(상품의 원래 가격), `bargainPrice`(상품의 할인된 가격), `stock`(잔여수량), `devidePrice`(상품 가격을 천의 단위로 , 넣어주는 함수)
<br>

### **View**
#### 1) LoadingIndicator
> 로딩화면을 시작하고 끝내는 클래스
- `showLoading`: 로딩화면을 실행하는 타입 메서드
- `hideLoading`: 로딩화면을 정지하는 타입 메서드

#### 2) MarketListCollectionViewCell
> 상품들을 List 형태로 보여주는 컬렉션 뷰 셀 
- `imageView`: 상품의 이미지
- `accessoryImageView`: 우측 `>` 이미지
- `nameLabel`: 상품의 이름 라벨
- `priceLabel`: 상품의 원가, 할인가격 라벨
- `stockLabel`: 상품의 잔여수량 라벨
- `horizontalStackView`: 상품 이름과 `subHorizontalStackView`를 묶어준 스택뷰
- `subHorizontalStackView`: 잔여수량, `>`를 묶어준 스택뷰
- `verticalStackView`: `horizontalStackView`와 상품 가격 라벨을 묶어준 스택뷰
- `entireStackView`: `verticalStackView`와 상품의 이미지를 묶어준 스택뷰
- `arrangeSubView`: 각 스택뷰의 Constraint를 잡아준 함수
- `extension CALayer {func addBottomBorder()}`: 셀 밑에 선처럼 보이게 `bottomBorder`를 만들어주는 함수


#### 3) MarketGridCollectionViewCell
> 상품들을 Grid 형태로 보여주는 컬렉션 뷰 셀
- `imageView`: 상품의 이미지
- `nameLabel`: 상품의 이름 라벨
- `priceLabel`: 상품의 원가, 할인가격 라벨
- `stockLabel`: 상품의 잔여수량 라벨
- `verticalStackView`: 위에 4가지 라벨을 묶어준 스택뷰
- `arrangeSubView`: 스택뷰의 Constraint를 잡아준 함수


### **ViewController**
#### 1) **MarketCollectionViewController**
> 상품 목록을 List와 Grid 형식으로 보여주는 컬렉션 뷰 셀을 관리하는 뷰컨트롤러
- `createListLayout` : List 형식의 `UICollectionViewLayout`을 반환하는 메서드
- `createGridLayout` : Grid 형식의 `UICollectionViewLayout`을 반환하는 메서드
- `makeListDataSource` : List Cell의 **Registration**을 설정하고 **DataSource**를 생성하는 메서드
- `makeGridDataSource` : : Grid Cell의 **Registration**을 설정하고 **DataSource**를 생성하는 메서드
- `receivePageData` : 서버를 통해 상품 데이터를 받아오는 메서드
- `applySnapshots` : 받아온 데이터로 **SnapShot**을 생성해 데이터소스에 적용하는 메서드
- `indexChanged` :`segmentedControl`의 값이 변경되었을 때 뷰의 모양을 바꾸는 메서드

<br>

---

## ⁉️ **고민 및 궁금한 점**
### 1. 뷰 Constraint 관련
#### 1) `contentView` vs `self`
```swift
// MarketGridCollectionViewCell.swift
NSLayoutConstraint.activate([
	 verticalStackView.topAnchor.constraint(equalTo: self.topAnchor, constant: 8),
         verticalStackView.bottomAnchor.constraint(equalTo: self.bottomAnchor, constant: -8),
	 verticalStackView.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -8),
	 verticalStackView.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 8),
	 imageView.heightAnchor.constraint(equalTo: self.heightAnchor, multiplier: 0.58)
])
```
Grid 셀과 List 셀 모두 콘텐츠에 대한 제약을 `self`와 관련하여 잡아주었습니다. **CollectionViewCell**에서 적어준 코드이니 `self`는 셀 자신이고 `contentView`는 셀 내부의 콘텐츠 영역이라고 생각해서 처음에는 `contentView`로 제약을 걸어주었습니다.

<img src="https://i.imgur.com/NFCiDqc.jpg" width="250" height="500" />|
<br>하지만 `contentView`에 제약을 설정하니 위 사진과 같이 콘텐츠(특히 이미지)가 셀 크기를 벗어나는 문제가 있어 `self`로 변경하였습니다.
왜 `contenView`일 때는 위와 같은 문제가 있던 것일까요? 그리고 지금처럼 `self`로 셀에 직접 제약을 걸어도 괜찮은지 궁금합니다!


#### 2) translatesAutoresizingMaskIntoConstraints 범위

`translatesAutoresizingMaskIntoConstraints = false` 는 전체를 감싼 `stackView`에서만 해주어도 되는지, 혹은 `stackView` 내부에 있는 모든 콘텐츠(label, image 등)마다 따로 설정을 해줘야하는건지 궁금합니다.

<br>

### 2. 콘솔 에러
처음 실행했을 때는 오류가 없다가 리스트 뷰에서 그리드 뷰로 전환하면 콘솔창에 `LayoutConstraints`관련 에러 메세지가 엄청 뜨는데, 빌드할 때는 지장이 가지 않고 뷰도 잘 나옵니다. 현재 `LayoutConstraints`에 문제가 있는 것일까요? 🥺

<img src="https://i.imgur.com/YElcx47.png" width="600" height="500" />|
<br>또한 항상 뜨는 `boringssl`이 어떤 것인지, 어떻게 처리를 해주어야하는지 감이 오지 않아 콘에게 여쭤봅니다 🥲

![](https://i.imgur.com/HdNO5w9.png)

<br>

### 3. 보라색 전구 표시
- 보라색 느낌표로 뜨는 에러는 **Cell**이나 **stackView**의 conflict로 생기는 에러로 알고 있어 수정해주었지만, 보라색 전구로 뜨는 것은 저희에게 다른 방식으로 제안을 하는 것인지 고치라고 하는 것인지 확실하지가 않아 따로 처리를 해주지 않았습니다. 이러한 경우에 참고만 하고 넘어가도 되는 것인지 고민이였습니다.

![](https://i.imgur.com/glJmUSM.png)

<br>

### 4. StackView에 ImageView 넣기
컬렉션 뷰를 표시하면서 가장 문제였던 것은 이미지입니다.. List와 Grid 형식 모두 먼저 크게 하나의 **StackView**로 **Cell**에 제약을 걸고 그 **StackView**에 **ImageView**와 **Label**을 넣었습니다. 하지만 특히 이미지가 셀을 벗어나거나 크기가 제멋대로 나오는 문제가 자주 발생했습니다. 현재는 이미지에 크기를 지정하여 표시되는 것은 해결하였는데 혹시 **StackView** 안에 **ImageView**를 넣는 것이 좋지 않은 방법일까요? 

| **제각각인 이미지 사이즈 예시** | **이미지 사이즈 제약 실패 예시** |
|:---:|:---:|
|<img src="https://i.imgur.com/yLg1jSC.jpg" width="300" height="600" />|![](https://i.imgur.com/xvb55xa.gif)|

<br>

## ⌨️ 커밋 규칙
* feat    : 기능 추가 (새로운 기능)
* fix      : 버그 수정
* refactor : 리팩토링 (네이밍 수정 등)
* style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
* docs    : 문서 변경 (문서 추가, 수정, 삭제)
* test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)

## 💻 코드 컨벤션
Swift Statements 형식 준수

ZZBAE committed:
feat: UICollectionView에서 사용할 Item 타입 구현

- step1의 UnitTest에서 unexpected error 원인이였던 creatAt과 issuedAt 타입 Date에서 String으로 변경

Judy-999 committed:
feat: MarketCollectionView의 DataSource 생성

- productCellRegistration() : 셀의 registration을 지정하는 메서드
- makeDataSource() : DataSource를 반환하는 메서드
- MarketCollectionView의 Section 추가

ZZBAE committed:
refactor: 그룹과 파일 위치 정리

- Utility, Session, ViewController 으로 그룹 분리 후 파일 위치 수정

ZZBAE committed:
feat: View의 레이아웃을 생성하는 메서드 구현

ZZBAE committed:
feat: 서버에서 받아온 데이터의 처리 과정 추가

- ViewController에 receivePageData 메서드가 페이지의 데이터를 가져올 수 있게 수정
-  Item의 init() 추가
- receivePage를 receiveData로 네이밍 수정

Judy-999 committed:
feat:  SnapShot 관련 메서드인 applySnapshots 추가

- Item의 init에서 하던 UIImage 변환을 MarketCollectionViewController에서 실행
- URLSessionManager의 baseURL을 제거하고 전체 URL을 받도록 함

Judy-999 committed:
feat: 커스텀 셀인 MarketListCollectionViewCell 추가

- createLayout 로직 변경

ZZBAE committed:
feat: Cell에 bottomBoarder 지정

- 각 cell에 bottomBoader를 cgcolor systemgray3로 설정해주어 cell을 구분해줌

ZZBAE committed:
feat: MarketGridCollectionViewCell 생성 및 layout 지정

- MarketGridCollectionViewCell의 layout을 잡아줌
- MarketCollectionViewController에서 GridLayout 메서드 생성 및 사이즈 지정
- makeGridDataSource 함수 구현 (기존 makeDataSource -> makeListDataSource 수정)
- 필요없는 주석 제거

Judy-999 committed:
feat: Grid cell의 border 지정

- Grid cell의 label font 지정

Judy-999 committed:
feat: segmentedControl 생성 및 기능 구현

ZZBAE committed:
feat: Grid 뷰 부분 수정

- price 와 bargainPrice의 label 통합 및 색깔 지정, currency 추가
- cell이 reuse 될 때 바뀌는 stockLabel과 priceLabel의 색깔 수정
- item의 잔여수량과 할인여부에 따른 label의 text 설정

Judy-999 committed:
feat: List cell에 stockLabel과 accessaryImageView 추가

- Grid cell의 이미지 사이즈 변경
- List cell과 Grid cell의 height 변경

ZZBAE committed:
feat: priceLabel 과 barbutton 구현 추가

- price에서 1000의 자리 기준으로 나눠주는 " , " 추가
- 네비게이션 오른쪽 상단에 바 버튼 + 추가

ZZBAE committed:
feat: LoadingIndicator 생성

- 처음 page data를 받아오기 전에 뜨는 로딩화면을 구현

Judy-999 committed:
refactor: 전체적으로 접근제어자(private) 추가

Judy-999 committed:
refactor: List cell의 뷰 관련 에러 수정

- accessaryImageView의 width를 지정

protocorn93 commented on OpenMarket/OpenMarket/Models/Item.swift:
```diff
@@ -0,0 +1,32 @@
+
+import UIKit
+
+struct Item: Hashable {
```
`Hashable`은 무엇을 위해 채택을 하셨나요? `Hashable`에 대해 설명부탁드려요~

Judy-999 replied:
`Diffable Data Source`는 `identifier`가 포함된 데이터 소스를 이용하여 컬렉션 뷰의 데이터 표시와 업데이트를 간소할 수 있다는 장점이 있습니다. `DataSource`와 달리 이전과 달라진 부분을 찾아내 간단하게 업데이트 할 수 있는데 이전과 달라졌는지 비교할 수 있으려면 해당 데이터 타입이 `Hashable` 해야 하기 때문에 채택해주었습니다!

protocorn93 commented on OpenMarket/OpenMarket/Models/Item.swift:
```diff
@@ -0,0 +1,32 @@
+        formatter.numberStyle = .decimal
+        return formatter.string(for: self) ?? ""
+    }
+}
```
`Item`에서만 사용하는 것이라면 `fileprivate extension`을 사용해도 좋을 것 같아요. 그게 아니라 범용적으로 사용될 것 같으면 별도로 파일로 분리하는게 좋을 것 같아요

ZZBAE replied:
Item에서만 사용되기 때문에 `fileprivate extension`으로 변경해주었습니다 :)

protocorn93 commented on OpenMarket/OpenMarket/View/LoadingIndicator.swift:
```diff
@@ -0,0 +1,27 @@
+            loadingIndicatorView.color = .systemGray
+            superView.addSubview(loadingIndicatorView)
+            loadingIndicatorView.startAnimating()
+        }
```
`hideLoading`도 그렇고 `DispatchQueue.main.async`로 감싼 이유가 무엇일까요?

ZZBAE replied:
`LoadingIndicator`를 뷰에서 나타나게 해주고 안보이게끔 해주는 UI 작업이니 메인 스레드에서 비동기작업을 해줘야한다고 생각해서 감싸주었는데,
벗겨내고 실행을 해도 빌드가 되어서 필요한 부분이 아니구나 생각이 들어 삭제해주었습니다.
저희가 DispatchQueue에 대한 이해나 접근을 잘못하고 있는걸까요?

protocorn93 replied:
의도는 맞지만 저는 보통 요런건 호출하는 쪽에서 결정해주면 좋을 것 같아요. 네트워크 요청 completion handler에서 해준다던지! 뭐 사실 크게 다르진 않긴해요! 👍

protocorn93 commented on OpenMarket/OpenMarket/View/LoadingIndicator.swift:
```diff
@@ -0,0 +1,27 @@
+import UIKit
+
+class LoadingIndicator {
+    static func showLoading(superView: UIView) {
```
## 스타일

`showloading(on view: UIView)`와 같은 형태의 아규먼트 레이블도 괜찮을 것 같네요!

Judy-999 replied:
`showloading`과 `hideLoading` 모두 `(on view: UIView)`로 변경했습니다!

protocorn93 commented on OpenMarket/OpenMarket/View/MarketGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,82 @@
+    let imageView: UIImageView = {
+        let imageView = UIImageView()
+        return imageView
+    }()
```
셀 내부의 자식 뷰들도 접근 제어자 붙여주시면 좋을 것 같아요.

ZZBAE replied:
말씀해주신대로 private이 붙지 않은 셀 내부의 자식 뷰들에 접근 제어자를 붙여주었습니다!

protocorn93 commented on OpenMarket/OpenMarket/View/MarketGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,82 @@
+        
+        self.addSubview(verticalStackView)
+        
+        self.layer.cornerRadius = 10
```
현재는 고정값이라 여기에 위치해도 무방하지만 레이아웃에 비례한 값을 지정할 경우에는 레이아웃이 모두 확정된 뒤에 값을 지정해야 원하는 모습을 그릴 수 있어요. 이점 기억해두시면 좋을 것 같아요!

ZZBAE replied:
고정값일 때는 위치가 상관없지만, 레이아웃에 비례할 때에는 모든 레이아웃을 설정해주고난 뒤에 지정해야 적용이 된다는 말씀이신거죠?!
꼭 기억해두겠습니다!

protocorn93 commented on OpenMarket/OpenMarket/View/MarketGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,82 @@
+        super.init(coder: coder)
+    }
+    
+    override func prepareForReuse() {
+        stockLabel.textColor = .systemGray
+        priceLabel.textColor = .systemRed
+    }
```
`prepareForReuse`를 오버라이드한 이유가 무엇인가요? 이 메서드의 목적은 무엇일까요?

ZZBAE replied:
`prepareForReuse`를 사용하기 전에는 원가, 할인된 가격라벨의 색깔이나 재고 수량의 품절 라벨이 뜬 후에
재사용된 셀의 글자 색깔이 오렌지에서 회색으로, 회색에서 빨강으로 돌아오지 않았습니다.
그래서 `prepareForReuse` 메서드를 사용하여 셀을 재사용하기 전에 초기화 작업을 해주고
`priceLabel` 과 `stockLabel`의 텍스트 색깔을 기존에 설정한 색깔로 나올 수 있게 해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/View/MarketGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,82 @@
+        ])
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        arrangeSubView()
+    }
+    
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+    }
```
`UITableViewCell`과 `UIColletionViewCell`의 메서드중엔 `awakeFromNib`이란 메서드도 있는데요. 

`init(frame: )`, `init?(coder: )`,` awakeFromNib` 각각에 대해 간략하게 설명해주실 수 있으신가요?

Judy-999 replied:
1) `init(frame: CGRect)`
- 코드로 뷰 개체를 만드는 경우 `init(frame: CGRect)`를 통해 초기화
- **Interface Builder**를 통해 디자인한 경우에는 **nib 파일**에서 뷰가 로드될 때 이 메서드는 호출되지 않는다

2) `init?(coder: NSCoder)`
- **스토리보드**나 **nib 파일**을 사용해 뷰를 커스텀했을 때 초기화해주는 메서드
- `UIView`가 `NSCoding`을 채택하고 있기 때문에 `UIView`를 상속받은 클래스에서도 `init?(coder: NSCoder)`를 해줘야 한다

3) `awakeFromNib`
- **Interface Builder**에서 객체가 `init(coder:)`로 초기화가 완료된 후 호출되는 메서드
- 객체가 `awakeFromNib` 메시지를 받으면 모든 outlet 및 action 연결이 이미 설정되어 있음이 보장된다
- 즉, 원래 `awakeFromNib()` 메서드가 호출된 시점에는 자신이 가진 IBOutlet들이 모두 초기화된 후 연결되어 접근이 가능한 상태여야 합니다

어려운 개념이라 제대로 이해하지 못한 부분이 있을 수 있지만 간략하게 `init(frame: CGRect)`는 코드로 작성된 뷰를 초기화하고, `init?(coder: NSCoder)`는 스토리보드를 통한 뷰를 초기화하는데 `UIView`를 상속했을 때 반드시 구현해야 하고, `awakeFromNib`는 `init?(coder: NSCoder)` 이후에 호출되는 메서드로 이해했습니다! 혹시 다른 점이 있다면 알려주시면 감사하겠습니다 🙇‍♀️

protocorn93 commented on OpenMarket/OpenMarket/View/MarketListCollectionViewCell.swift:
```diff
@@ -0,0 +1,133 @@
+    let imageView: UIImageView = {
+        let imageView = UIImageView()
+        return imageView
+    }()
```
마찬가지로 접근제어자 명시해주세요~

Judy-999 replied:
넵 명시해줬습니다!

protocorn93 commented on OpenMarket/OpenMarket/View/MarketListCollectionViewCell.swift:
```diff
@@ -0,0 +1,133 @@
+
+import UIKit
+
+class MarketListCollectionViewCell: UICollectionViewCell {
```
`final` 키워드를 붙이는 습관을 가져가는 것이 좋아요!

Judy-999 replied:
넵! 상속하지 않을 class에 모두 `final` 키워드를 붙여줬습니다.

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        let addButton = UIBarButtonItem()
+        addButton.image = UIImage(systemName: "plus")
+        return addButton
+    }()
```
## 스타일

이건 개인 스타일 차이일 수 있으나 `// MARK: `를 이용해서 코드의 물리적인 구조를 나누고 설명을 붙이는 것은 코드를 읽고 찾는데 도움이 되어요. 

저 같으면 이렇게 분류할 것 같아요. 

```swift 
// MARK: Inner types 

...

// MARK: UI

...

// MARK: Properties 

...
```

ZZBAE replied:
추천해주신 방식대로 `// MARK: `를 사용해 ViewController에서 역할별로 구분을 해주었습니다. 분리가 되어 있으니까 확실히 깔끔해보이고 찾기도 쉬워서 좋은 것 같습니다 😀👍🏻

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        navigationItem.rightBarButtonItem = barbutton
+        collectionView.collectionViewLayout = createListLayout()
+        receivePageData()
+        segmentedControl.addTarget(self, action: #selector(indexChanged), for: .valueChanged)
```
코드의 추상화 레벨이 서로 달라 코드 가독성이 떨어지는 것 같아요. (코드 추상화 레벨 - 클린코드 참고) 

한 줄 짜리 메서드라도 의도를 잘 나타내는 네이밍의 메서드로 관리하는 것도 좋은 방법이에요. `configureUI`라는 네이밍의 메서드로 분리해볼 수 있는 것들이 보여요. 이외에도 여러분들의 기준으로 한번 분리해보세요

ZZBAE replied:
`segmentedControl`, `barbutton`, `createListLayout()` -> `configureUI()`
`segmentedControl.addTarget(self, action: #selector(indexChanged), for: .valueChanged)` ->`addAction()`

이렇게 메서드를 분리해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/View/MarketGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,82 @@
+        let label = UILabel()
+        label.font = .boldSystemFont(ofSize: 20)
+        return label
+    }()
```
`UIStackView`의 `arrangedSubviews`가 되는 뷰들은 내부적으로 `translatesAutoresizingMaskIntoConstraints`가 `false`가 되지만 그래도 명시해주는 것이 좋아요.

ZZBAE replied:
내부적으로 되어있긴 하지만 콘이 말씀해주신 것처럼 습관적으로 붙이기 위해 각 컨텐츠별로 `translatesAutoresizingMaskIntoConstraints = false`를 명시해주었습니다 ☺️

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -0,0 +1,218 @@
+            } else {
+                cell.stockLabel.text = "품절"
+                cell.stockLabel.textColor = .systemOrange
+            }
```
셀에는 데이터(Item)만 전달하고 내부에서 그리도록 해주세요. 데이터를 받아 화면에 그리는 형태로 가공하는 것은 셀의 역할이에요! `configure(item:)`과 같은 메서드만 외부로 공개하는게 좋아요

Judy-999 replied:
**MarketGridCollectionViewCell**과 **MarketListCollectionViewCell**에 `configureCell(with item: Item)` 메서드를 추가해 해당 메서드로 `item`만 전달해 cell을 그리도록 했습니다!

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -0,0 +1,218 @@
+ 
+extension MarketCollectionViewController {
+    private func applySnapshots() {
+        var itemSnapshot = NSDiffableDataSourceSnapshot<Section, Item>()
```
## 스타일

`typealias`로 DataSource를 별도 타입으로 정의한 것처럼 Snapshot도 할 수 있지 않을까요?

ZZBAE replied:
`typealias SnapShot = NSDiffableDataSourceSnapshot<Section, Item>` 이렇게 별도 타입으로 정의해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -0,0 +1,218 @@
+                    LoadingIndicator.hideLoading(superView: self.view)
+                }
+            case .failure(_):
+                print("서버 통신 실패")
```
알럿을 띄어 에러를 사용자에게 보여주는게 좋을 것 같아요.

ZZBAE replied:
서버 통신이 실패되어 데이터를 불러오지 못함을 알려주는 얼럿을 생성해주었습니다.
```
    private func showAlert() {
        let failureAlert = UIAlertController(title: "서버 통신 실패", message: "데이터를 받아오지 못했습니다.", preferredStyle: .alert)
        failureAlert.addAction(UIAlertAction(title: "확인", style: .default))
        present(failureAlert, animated: true)
    }
```

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        sessionManager.receiveData(baseURL: subURL) { result in
+            switch result {
+            case .success(let data):
+                guard let page = DataDecoder().decode(type: Page.self, data: data) else { return }
```
여기도 별도 에러 처리나 실패 처리가 필요해보여요

Judy-999 replied:
파싱할 때 발생할 수 있는 에러 타입으로 `ParsingError`를 생성하고  `DataDecoder`의 `decode`메서드가 에러를 `throw`하도록 했습니다.
**Page** 데이터를 decode하는 부분에서 에러가 발생하면 통신 실패와 마찬가지로 얼럿을 띄우도록 했습니다!

protocorn93 reviewed (changes requested):
두 분 쉽지 않은 미션인데 수고하셨어요! 👏

## 고민 및 궁금한 점

> contentView vs self

contentView에 해주어야 해요. [공식 문서](https://developer.apple.com/documentation/uikit/uitableviewcell/1623229-contentview)를 읽어보시면 

"you should add them to the content view so they position appropriately as the cell transitions in to and out of editing mode."

라고 설명하는데요. 여기서 editing mode란

<img width="779" alt="스크린샷 2022-07-22 오전 12 15 30" src="https://user-images.githubusercontent.com/20634806/180250062-ea42b7ce-d092-4559-8538-6da3d11d2dd6.png">

이렇게 UIKit이 제공하는 편집모드시 레이아웃이 변경이 될텐데, 이에 맞춰 적절하게 레이아웃이 배치되도록 하려면 contentView에 배치해야한다는 말이에요. 

저는 여러분 코드에 contentView로 제약을 주어도 잘 노출되더라구요

| code | grid | list |
:-:|:-:|:-:
| <img width="994" alt="content-view" src="https://user-images.githubusercontent.com/20634806/180250417-e551b769-db9d-4f22-a77b-685649a53ef9.png"> | ![Simulator Screen Shot - iPhone 12 Pro - 2022-07-22 at 00 16 45](https://user-images.githubusercontent.com/20634806/180250442-85ce8ded-f297-42bd-886f-38d15b80a343.png) | ![Simulator Screen Shot - iPhone 12 Pro - 2022-07-22 at 00 16 42](https://user-images.githubusercontent.com/20634806/180250438-857ac309-2b21-4379-8218-a7ce0bed6535.png) |

> translatesAutoresizingMaskIntoConstraints 범위

이건 코멘트에도 남겨놓았는데 `UIStackView`의 `arrangedSubviews`들은 내부적으로 자동으로 false로 되요. 그래도 뷰 정의시 false로 지정하는 습관을 가져가는게 좋아요!

> 레이아웃 콘솔 에러

마찬가지로 제가 여러분의 코드를 돌려보았을 땐 해당 콘솔 에러가 발생하지 않는데요. 해당 로그가 말하는 것은 동일한 제약이 걸려있고 그 중 어떤걸 선택할지 모르기 때문에 UIKit이 하나를 선택했다는 뜻이에요. 

중복으로 혹은 UIKit이 판단하기에 어떤걸 선택할지 헷갈리는 제약이 있는지 찾아보시면 되요. 로그 이미지가 잘린 것 같아서 제가 명확히 알 수가 없네요. 보통 결국 어느걸 선택했다고 나오거든요. 

https://www.wtfautolayout.com 이 사이트의 도움을 받아볼수도 있어요. 

그리고 boringssl 로그는 애플 이슈인것 같은데 저도 찾아봤는데 다들 원인은 못찾은 것 같아요. 이 글을 참고해보세요. https://developer.apple.com/forums/thread/697026


> 보라색 전구 표시

이 역시 저는 나타나지 않는데, 보통 layer에서 비용이 많이 드는 작업을 수행할 때 나타나요. 그래서 메시지가 하는 말이 그림자를 그릴거면 shadowPath를 사용하라는 뜻이에요. 

https://stackoverflow.com/questions/64277067/how-to-fix-optimization-opportunities 이 링크를 참고해보세요! 


> StackView에 ImageView 넣기

스택뷰에 이미지뷰를 당연히 넣을 수 있죠. 다만 이미지뷰에 별도 사이즈를 지정하거나 제약을 가하지 않으면 이미지 크기가 이미지뷰의 크기가 되어요 그래서 첨부해주신 사진 처럼 나오는거에요. 

이를 intrinsic content size라고 하는데 이에 대해 학습해보시는걸 권장해드려요. https://velog.io/@wonhee010/Intrinsic-Content-Size 이 글도 괜찮더라구요

ZZBAE committed:
refactor: constraints 및 접근제어자 수정

- List, Grid, LoadingIndicator class 앞에 final 추가
- 각 컨텐츠마다 translatesAutoresizingMaskIntoConstraints = false 추가
- List 와 Grid Cell에 constraints를 contentView에 제약

Judy-999 committed:
refactor: item만 전달해서 Cell 내부에서 그리도록 변경

- configureCell(with item: Item) 메서드 구현

Judy-999 committed:
refactor: 접근제어자 및 아규먼트 레이블 변경

- Int의 devidePrice() 메서드를 fileprivate으로 변경
- List와 Grid Cell 내부 뷰들을 private으로 변경
- LoadingIndicator의 showLoading과 hideLoading의 아규먼트 레이블을 (on view: UIView)로 변경

ZZBAE committed:
refactor: Mark로 코드구분 및 snapshot typealias 설정

- // Mark: 를 사용하여 역할 분리
- typealias로 snapshot을 별도의 타입으로 지정

ZZBAE committed:
refactor: viewDidLoad() 함수 분리 및 실패 얼럿 생성

- viewDidLoad() 의 함수 분리를 통한 추상화 레벨 통일
- 데이터 통신 실패의 경우 나타내주는 얼럿 생성

Judy-999 committed:
refactor: 데이터 파싱할 때 발생할 수 있는 ParsingError 생성

- Page 데이터 decode를 실패하면 얼럿을 띄우도록 설정

ZZBAE committed:
feat: README.md 추가

- 오픈마켓 step 1, 2의 리드미 파일 업데이트

Judy-999 commented:
리뷰 감사합니다 콘 😊
남겨주신 코멘트 읽고 리팩토링 후 댓글도 함께 남겨놓았습니다 😀

❗️ 고민 및 궁금한 점 해결
### 1. 뷰 Constraint 관련
`contenView`에 제약을 거는 것으로 변경해주었습니다.

이전에 `contentView`에 제약을 걸었을 때 문제가 생겼던 이유는 `self`에 `addsubView()`로 `stackView`를 추가한 채로 제약만 `contentView`에 걸었기 때문이었습니다..😅 `contentView`에  `addsubView()` 하는 것으로 변경하니 해결되었습니다.
<br>

### 2. 콘솔 에러
1번 내용인 셀의 아이템들을 `contentView`에 넣고 제약을 걸어주었더니 자연스럽게 해결된 것으로 보아 아마 `self`로 셀 자체에 제약을 건 상태로 `Layout`을 변경하면서 문제가 생겼던 것 같습니다!
